File: Objetos_2_java/src/main/java/objetos2/java/gestionBiblioteca/Biblioteca.java
package objetos2.java.gestionBiblioteca;

import java.util.ArrayList;

public class Biblioteca {

    private ArrayList<Libro> libros = new ArrayList<>();
    private ArrayList<Usuario> usuarios = new ArrayList<>();


    public void agregarLibro(Libro libro) {
        if (!libros.contains(libro)) {
            libros.add(libro);
            System.out.println("Libro " + libro.titulo + " agregado correctamente");
        }else{
            System.out.println("Este libro ya existe en la lista");
        }
    }

    public void registrarUsuario(Usuario usuario) {
        if (!usuarios.contains(usuario)) {
            usuarios.add(usuario);
            System.out.println("Usuario " + usuario.nombre + " agregado correctamente");
        }else{
            System.out.println("Este usuario ya existe en la lista");
        }
    }

    public void prestarLibro(Libro libro, Usuario usuario) {
        if (libro.prestado){
            System.out.println("Error. El libro ya esta prestado");
        }else{
            usuario.prestamos.add(libro);
            libro.prestado = true;
            System.out.println("Libro " + libro.titulo + " prestado correctamente");
        }
    }

    public void devolverLibro(Libro libro, Usuario usuario) {
        if (!libro.prestado){
            System.out.println("Error. El libro no está prestado");
        }else{
            if (usuario.prestamos.contains(libro)){
                usuario.prestamos.remove(libro);
                libro.prestado = false;
                System.out.println("Libro " + libro.titulo + " devuelto correctamente");
            }else{
                System.out.println("Error al devolver. El libro " + libro.titulo + " lo tiene otro usuario");
            }
        }
    }

    public void mostrarInfoUsersLibros(){
        System.out.println("Lista de usuario regsitrados: ");

        for (Usuario usuario : usuarios) {
            if (!usuario.prestamos.isEmpty()){
                System.out.println("- " + usuario.nombre + " tiene prestados: ");

                for (Libro libro : usuario.prestamos) {
                    libro.mostrarInfo();
                }
            }else{
                System.out.println("- " + usuario.nombre + ", sin prestamos");
            }
        }
    }


}

File: Objetos_2_Kotlin/src/main/kotlin/gestionBiblioteca/Biblioteca.kt
package gestionBiblioteca

class Biblioteca {

    private val libros = mutableListOf<Libro>()
    private val usuarios = mutableListOf<Usuario>()

    fun agregarLibro(libro: Libro) {
        if (!libros.contains(libro)) {
            libros.add(libro)
            println("Libro añadido correctamente")
        }else{
            println("Ya existe este libro.")
        }
    }

    fun registrarUsuario(usuario: Usuario) {
        if (!usuarios.contains(usuario)) {
            usuarios.add(usuario)
            println("Usuario registrado correctamente")
        }else{
            println("Ya existe este usuario.")
        }
    }

    fun prestarLibro(libro: Libro, usuario: Usuario) {
        if (libro.prestado){
            println("Error. El libro está actualmente prestado.")
        }else{
            usuario.prestamos.add(libro)
            libro.prestado = true
            println("Libro prestado correctamente.")
        }
    }

    fun devolverLibro(libro: Libro, usuario: Usuario) {
        if (!libro.prestado){
            println("El libro no está prestado")
        }else{
            if (usuario.prestamos.contains(libro)){
                usuario.prestamos.remove(libro)
                libro.prestado = false
                println("Libro devuelto correctamente")
            }else{
                println("Error al devolver. El libro lo tiene otro usuario!!")
            }
        }
    }

    fun mostrarInfoUsersLibros(){
        println("Lista usuarios registrados: ")

        usuarios.forEach {usuario ->
            if (usuario.prestamos.isNotEmpty()){
                println("${usuario.nombre} tiene prestados: ")
                usuario.prestamos.forEach {libro ->
                    println(libro)
                }
            }else{
                println("${usuario.nombre}, sin préstamos")
            }


        }
    }

}